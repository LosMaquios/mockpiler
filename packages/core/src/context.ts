export type MockContextInput = MockContext | MockContextAccessor

export interface MockContext {
  [key: string]: any
}

export type MockContextAccessor = (key: string) => any

export const unknownIdent = Symbol('MockPiler.UnknownIdent')

const encodeKey = (n: number) => n.toString(36)

class RawValue {
  constructor (
    public raw: any) {}
}

export function raw (value: any) {
  return new RawValue(value)
}

export function getTemplateAndRootContext (
  templateStrings: TemplateStringsArray, 
  values: any[],
  accessor: MockContextAccessor
): [string, MockContext] {
  let [template] = templateStrings
  const randomContextKey = encodeKey(Math.random()).slice(2)

  const rootContext: MockContext = {}
  const rootContextProxy = new Proxy(rootContext, {
    get (target, key: string) {
      return Reflect.has(target, key)
        ? target[key]
        : accessor(key)
    }
  })

  values.forEach((value, index) => {
    if (value instanceof RawValue) {
      template += `${value.raw}${templateStrings[index + 1]}`
    } else {
      const contextKey = `__mockpiler__${randomContextKey}_${encodeKey(index)}`

      rootContext[contextKey] = value
      template += `'${contextKey}'${templateStrings[index + 1]}`
    }
  })

  return [
    template,
    rootContextProxy
  ]
}

/**
 * Utility function to get a context accessor 
 * from a plain context object
 * 
 * @param input
 */
export function getContextAccessor (input: MockContextInput): MockContextAccessor {
  if (typeof input === 'function') {
    return input as any
  }

  // Generate an accessor based on given input
  return key => key in input ? input[key] : unknownIdent
}

/**
 * Merges given plain context objects or accessors
 * 
 * @param inputs 
 */
export function mergeContexts (...inputs: MockContextInput[]): MockContextAccessor {
  const accessors = inputs.map(getContextAccessor)

  return key => {
    for (const accessor of accessors) {
      const result = accessor(key)

      if (result !== unknownIdent) {
        return result
      }
    }

    return unknownIdent
  }
}
