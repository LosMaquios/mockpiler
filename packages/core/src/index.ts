export * from './parser'
export * from './lexer'
export { 
  createCompiler as default,
  raw
} from './compiler'
