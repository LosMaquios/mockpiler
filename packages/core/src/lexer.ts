export enum TokenChar {
  /**
   * Array tokens
   */
  arrayStartToken = '[',
  arrayEndToken = ']',

  /**
   * Object tokens
   */
  objectStartToken = '{',
  objectEndToken = '}',
  objectPairSeparator = ':',

  /**
   * Count tokens
   */
  countStartToken = '(',
  countEndToken = ')',

  /**
   * Identifier tokens
   */
  identifierToken = '\'',

  /**
   * Escape tokens
   */
  escapeToken = '\\'
}

export enum TokenType {
  identifier = 'identifier',
  object = 'object',
  array = 'array',
  count = 'count',
  countNumber = 'countNumber',
  EOF = 'EOF'
}

export interface Token {
  type: TokenType
  value: number | string
  location: {
    start: TokenLocation
    end: TokenLocation
  }
}

export interface TokenArray extends Token {
  type: TokenType.array
}

export interface TokenLocation {
  line: number
  column: number
}

/**
 * Valid start identifier chars are:
 * 
 *   1. An underscore _
 *   2. Any uppercase or lowercase letter
 */
export const START_IDENTIFIER_REGEX = /[a-z_]/i

/**
 * Valid identifiers chars remain are:
 * 
 *   1. Hyphens -
 *   2. Dots .
 *   3. Underscores _
 *   4. Any uppercase or lowercase letters
 */
export const IDENTIFIER_REGEX = /[-._a-z]/i

const COUNT_DIGIT_REGEX = /\d/
const LINE_CHAR = '\n'
const TAB_CHAR = '\t'
const SPACE_CHAR = ' '

/**
 * Chars to be ignored during scanning
 */
const IGNORED_CHARS = [
  '\r'
]

/**
 * Count of spaces by tab char
 */
const TAB_SPACE_SIZE = 4

const ARRAY_TOKENS = [
  TokenChar.arrayStartToken,
  TokenChar.arrayEndToken
]

const OBJECT_TOKENS = [
  TokenChar.objectStartToken,
  TokenChar.objectEndToken,
  TokenChar.objectPairSeparator
]

const COUNT_TOKENS = [
  TokenChar.countStartToken,
  TokenChar.countEndToken
]

export class LexerError extends Error {
  name = 'LexerError'
}

export function scan (input: string): Token[] {
  let index = 0
  let line = 1
  let column = 1

  const tokens: Token[] = []
  const { length } = input

  while (index < length) {
    if (is(ARRAY_TOKENS)) {
      tokens.push(consumeToken(TokenType.array))
    } else if (is(OBJECT_TOKENS)) {
      tokens.push(consumeToken(TokenType.object))
    } else if (is(COUNT_TOKENS)) {
      tokens.push(consumeToken(TokenType.count))
    } else if (is(COUNT_DIGIT_REGEX)) {
      const token = consumeRegex(TokenType.countNumber, COUNT_DIGIT_REGEX)
      token.value = parseInt(token.value as string)

      tokens.push(token)
    } else if (is(TokenChar.identifierToken)) {
      tokens.push(consumeIdent())
    } else if (is(START_IDENTIFIER_REGEX)) {
      tokens.push(consumeRegex(TokenType.identifier, IDENTIFIER_REGEX))
    } else if (is(LINE_CHAR)) {
      ++line
      ++index

      // Reset column
      column = 1
    } else if (is(TAB_CHAR)) {
      ++index
      column += TAB_SPACE_SIZE
    } else if (is(SPACE_CHAR)) {
      advance()
    } else if (is(IGNORED_CHARS)) {
      ++index
    } else {
      throw new LexerError(`Unknown token '${peek()}' at ${line}:${column}`)
    }
  }

  // Push end-of-file
  tokens.push({
    type: TokenType.EOF,
    value: '',
    location: {
      start: getLocation(),
      end: getLocation()
    }
  })

  return tokens

  function consumeToken (type: TokenType): Token {
    return {
      type,
      value: peek(),
      location: {
        start: getLocation(),
        end: (advance(), getLocation())
      }
    }
  }

  function consumeIdent (): Token {
    let ident = peek() // Save leading identifier token
    let escaping = false

    const start = getLocation()

    while (
      advance() &&
      (!is(TokenChar.identifierToken) || escaping)
    ) {
      escaping = !escaping && is(TokenChar.escapeToken)

      // Avoid appending escaping char
      if (!escaping) {
        ident += peek()
      }
    }

    if (!peek()) {
      throw new LexerError('Unexpected EOF')
    }

    // Save trailing identifier token
    ident += peek()

    // Skip trailing identifier token
    advance()

    return {
      type: TokenType.identifier,
      value: ident,
      location: {
        start,
        end: getLocation()
      }
    }
  }

  function consumeRegex (type: TokenType, regex: RegExp): Token {
    let value = peek()
    const start = getLocation()

    while (advance() && regex.test(peek())) {
      value += peek()
    }

    return {
      type,
      value,
      location: {
        start,
        end: getLocation()
      }
    }
  }

  function getLocation (): TokenLocation {
    return {
      line,
      column
    }
  }

  function is (expectation: RegExp | string | string[]) {
    return typeof expectation === 'string'
      ? peek() === expectation
      : Array.isArray(expectation)
      ? expectation.indexOf(peek() as any) > -1
      : expectation.test(peek())
  }

  function advance () {
    ++column
    ++index

    return peek()
  }

  function peek () {
    return input[index]
  }
}
