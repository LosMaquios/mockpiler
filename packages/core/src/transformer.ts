const gl = typeof window !== 'undefined'
  ? window
  : global

export type TransformFn = (value: any) => any

const Map: TransformFn = value => {
  if (typeof value !== 'object' || value === null) {
    throw new TypeError('Invalid value')
  }

  const map = new gl.Map<any, any>()

  if (!Array.isArray(value)) {
    const keys = Object.keys(value)

    for (const key of keys) {
      map.set(key, value[key])
    }
  } else {
    for (const element of value) {
      if (!Array.isArray(element) || element.length !== 2) {
        throw new TypeError('Expecting an entry array: [key, value]')
      }

      map.set(...element as [any, any])
    }
  }

  return map
}

const Set: TransformFn = value => {
  if (!Array.isArray(value)) {
    throw new TypeError('Invalid value')
  }

  return new gl.Set<any>(value)
}

export const transformers = {
  Map,
  Set
}