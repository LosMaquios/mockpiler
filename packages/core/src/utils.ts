type CloneFn = <T extends object>(obj: T) => T

export const shallowClone: CloneFn = obj => ({
  ...obj
})

export const deepClone: CloneFn = obj => (
  JSON.parse(
    JSON.stringify(
      obj
    )
  )
)
