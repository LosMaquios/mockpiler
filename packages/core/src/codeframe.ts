import { Token } from './lexer'

const PADDING_CHAR = ' '
const LINE_SEPARATOR = '|'

/**
 * Indicator for code fragments:
 * 
 *   wrong-fragment
 *   ^^^^^^^^^^^^^^
 */
const FRAGMENT_INDICATOR_CHAR = '^'

/**
 * Indicator for lines:
 * 
 *  > 1|
 *    2|
 *    3|
 */
const LINE_INDICATOR_CHAR = '>'

/**
 * Max lines above and below the highlighted line
 */
const MAX_LINES = 4

class CodeframeBuffer {
  private linesBuffer: { [line: number]: string } = {}

  constructor (
    private minLine: number,
    private maxLine: number) {}

  hasLine (line: number) {
    return line in this.linesBuffer
  }

  addLine (line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('')
  }

  prependToLine (line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('') + this.linesBuffer[line]
  }

  appendToLine (line: number, ...values: any[]) {
    this.linesBuffer[line] += values.join('')
  }

  getFormattedOutput (highlightLocation: Token['location']) {
    const formattedCode = []
    const maxLineLength = String(this.maxLine).length
    const lineIndicatorSpacing = LINE_INDICATOR_CHAR.length + 1

    for (let line = this.minLine; line <= this.maxLine; line++) {
      const inHighlightLine = line === highlightLocation.start.line

      formattedCode.push(
        (inHighlightLine ? LINE_INDICATOR_CHAR : '') +
        getPadding((inHighlightLine ? LINE_INDICATOR_CHAR.length : lineIndicatorSpacing) + maxLineLength - String(line).length) +
        line + LINE_SEPARATOR +
        (this.linesBuffer[line] ?? '')
      )

      if (inHighlightLine) {
        formattedCode.push(
          getPadding(maxLineLength + lineIndicatorSpacing) +
          LINE_SEPARATOR + 
          getPadding(highlightLocation.start.column - 1) +
          getFragmentIndicator(highlightLocation.end.column - highlightLocation.start.column)
        )
      }
    }

    return formattedCode.join('\n')
  }
}

export function generateCodeframe (
  tokens: Token[], 
  highlightTokenIndex: number
) {
  const { location: highlightLocation, value: highlightValue } = tokens[highlightTokenIndex]
  const MIN_LINE = Math.max(highlightLocation.start.line - MAX_LINES, 1)
  const MAX_LINE = Math.min(
    highlightLocation.start.line + MAX_LINES, 
    tokens[tokens.length - 1].location.start.line
  )

  const buffer = new CodeframeBuffer(MIN_LINE, MAX_LINE)

  buffer.addLine(highlightLocation.start.line, highlightValue)

  let upperIndex = highlightTokenIndex - 1
  let lowerIndex = highlightTokenIndex + 1

  let upperToken: Token
  let lowerToken: Token

  let prevUpperTokenStartLocation = highlightLocation.start
  let prevLowerTokenEndLocation = highlightLocation.end

  /**
   * Build upper code block
   */
  while (
    upperIndex >= 0 &&
    (upperToken = tokens[upperIndex--]).location.start.line >= MIN_LINE
  ) {
    const { value, location: { start, end } } = upperToken

    if (buffer.hasLine(start.line)) {
      buffer.prependToLine(
        start.line, 
        value, 
        getPadding(prevUpperTokenStartLocation.column - end.column)
      )
    } else {
      buffer.addLine(start.line, value)
      buffer.prependToLine(
        prevUpperTokenStartLocation.line, 
        getPadding(prevUpperTokenStartLocation.column - 1)
      )
    }

    prevUpperTokenStartLocation = start
  }

  buffer.prependToLine(
    prevUpperTokenStartLocation.line, 
    getPadding(prevUpperTokenStartLocation.column - 1)
  )

  /**
   * Build lower code block
   */
  while (
    lowerIndex < tokens.length &&
    (lowerToken = tokens[lowerIndex++]).location.start.line <= MAX_LINE
  ) {
    const { value, location: { start, end } } = lowerToken

    if (!buffer.hasLine(start.line)) {
      buffer.addLine(start.line, getPadding(start.column - 1), value)
    } else {
      buffer.appendToLine(
        start.line, 
        getPadding(start.column - prevLowerTokenEndLocation.column), 
        value
      )
    }

    prevLowerTokenEndLocation = end
  }

  return buffer.getFormattedOutput(highlightLocation)
}

function getFragmentIndicator (size: number) {
  return FRAGMENT_INDICATOR_CHAR.repeat(size)
}

function getPadding (size: number) {
  return PADDING_CHAR.repeat(size)
}
