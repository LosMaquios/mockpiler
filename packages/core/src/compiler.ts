import { scan } from './lexer'
import {
  parse,
  AstArrayNode,
  AstElementNode,
  AstValueNode,
  AstIdentifierNode,
  AstObjectNode,
  AstRootNode,
  AstNodeType,
  AstSpreadNode
} from './parser'
import {
  MockContextInput,
  MockContext,
  getTemplateAndRootContext,
  getContextAccessor,
  unknownIdent
} from './context'

export type CompileMockArgs = Parameters<typeof String['raw']>

export class CompilerError extends Error {
  name = 'CompilerError'
}

const EMPTY_CONTEXT = Object.create(null)

export function createCompiler (contextOrAccessor: MockContextInput): (...args: CompileMockArgs) => object
export function createCompiler (...args: CompileMockArgs): object
export function createCompiler (input: MockContextInput | TemplateStringsArray, ...values: any[]) {
  return Array.isArray(input) && input.hasOwnProperty('raw') 
    ? compileMockWithContext(input as any, values, EMPTY_CONTEXT)
    : (...[templateStrings, ...values]: CompileMockArgs) => compileMockWithContext(
        templateStrings,
        values,
        input
      )

  function compileMockWithContext (
    templateStrings: TemplateStringsArray,
    values: any[],
    contextOrAccessor: MockContextInput
  ) {
    return compileMock(
      ...getTemplateAndRootContext(
        templateStrings,
        values,
        getContextAccessor(contextOrAccessor)
      )
    )
  }
}

function compileMock (input: string, context: MockContext) {
  const tokens = scan(input)
  const rootNode = parse(tokens)

  return compileRoot(rootNode)

  function compileRoot ({ value }: AstRootNode) {
    return value.type === AstNodeType.Array
      ? compileArray(value)
      : compileObject(value)
  }

  function compileArray (node: AstArrayNode) {
    const result = []

    for (const element of node.elements) {
      if (element.type === AstNodeType.Spread) {
        compileElementSpread(element, result)
      } else {
        compileElement(element, result)
      }
    }

    return result
  }

  function compileElementSpread (node: AstSpreadNode, values: any[]) {
    const result = compileIdent(node.identifier)

    for (const value of result) {
      values.push(value)
    }
  }

  function compileElement (node: AstElementNode, values: any[]) {
    let count = typeof node.count !== 'number'
      ? compileIdent(node.count)
      : node.count

    while (count--) {
      values.push(
        compileValue(node.value)
      )
    }
  }

  function compileObject (node: AstObjectNode) {
    const result = {}

    for (const property of node.properties) {
      if (property.type === AstNodeType.Spread) {
        compilePropertySpread(property, result)
      } else {
        result[property.key.name] = compileValue(property.value)
      }
    }

    return result
  }

  function compilePropertySpread (node: AstSpreadNode, values: object) {
    const result = compileIdent(node.identifier)

    for (const key of Object.keys(result)) {
      values[key] = result[key]
    }
  }

  function compileValue (node: AstValueNode) {
    switch (node.type) {
      case AstNodeType.Array:
        return compileArray(node)

      case AstNodeType.Object:
        return compileObject(node)

      case AstNodeType.Identifier:
        return compileIdent(node)
    }
  }

  function compileIdent ({ name }: AstIdentifierNode) {
    const value = context[name]

    if (value === unknownIdent) {
      throw new CompilerError(`Unknown context identifier: ${name}`)
    }

    return typeof value === 'function'
      ? value.call(context)
      : value
  }
}