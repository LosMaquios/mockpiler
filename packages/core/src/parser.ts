import {
  Token,
  TokenChar,
  TokenType,
  TokenLocation
} from './lexer'
import { generateCodeframe } from './codeframe'
import {
  deepClone,
  shallowClone
} from './utils'

export enum AstNodeType {
  Root = 'Root',
  Identifier = 'Identifier',
  Spread = 'Spread',
  Transform = 'Tranform',
  Object = 'Object',
  Array = 'Array',
  Property = 'Property',
  Element = 'Element'
}

export interface AstNode {
  type: AstNodeType
  location: Token['location']
}

export interface AstPropertyNode extends AstNode {
  type: AstNodeType.Property
  key: AstIdentifierNode
  value: AstValueNode
}

export interface AstIdentifierNode extends AstNode {
  type: AstNodeType.Identifier
  name: string
}

export interface AstArrayNode extends AstNode {
  type: AstNodeType.Array
  elements: Array<AstSpreadNode | AstElementNode>
}

export interface AstElementNode extends AstNode {
  type: AstNodeType.Element
  count: number | AstIdentifierNode
  value: AstValueNode
}

export interface AstObjectNode extends AstNode {
  type: AstNodeType.Object
  properties: Array<AstSpreadNode | AstPropertyNode>
}

export interface AstRootNode extends AstNode {
  type: AstNodeType.Root
  value: AstObjectOrArrayNode
}

export interface AstSpreadNode extends AstNode {
  type: AstNodeType.Spread
  identifier: AstIdentifierNode
}

export interface AstTransformNode extends AstNode {
  type: AstNodeType.Transform
  transformer: AstIdentifierNode
  value: AstValueNode
}

type AstObjectOrArrayNode = AstObjectNode | AstArrayNode
export type AstValueNode = AstObjectOrArrayNode | AstIdentifierNode | AstTransformNode

const TRIM_IDENTIFIER_REGEX = new RegExp(`^${TokenChar.identifierToken}|${TokenChar.identifierToken}$`, 'g')

export class ParserError extends Error {
  name = 'ParserError'
}

export class Parser {
  index = 0

  constructor (
    public tokens: Token[]) {}

  parse () {
    return this.parseRoot()
  }

  parseRoot (): AstRootNode {
    const value: AstRootNode['value'] = (
      this.parseArray() ??
      this.parseObject()
    )

    if (!value) {
      this.throwUnexpected([
        TokenChar.arrayStartToken,
        TokenChar.objectStartToken
      ])
    }

    return {
      type: AstNodeType.Root,
      value,
      location: {
        start: { line: 1, column: 1 },
        end: shallowClone(this.current().location.end)
      }
    }
  }

  parseArray (): AstArrayNode {
    if (!this.is(TokenChar.arrayStartToken)) {
      return null
    }

    const startLocation: TokenLocation = shallowClone(this.current().location.start)
    const elements: AstArrayNode['elements'] = []

    // Skip start token `[`
    this.next()

    while (this.current() && !this.is(TokenChar.arrayEndToken)) {
      elements.push(
        this.parseSpread() ??
        this.parseElement()
      )
    }

    this.expect(TokenChar.arrayEndToken)

    const endLocation = shallowClone(this.current().location.end)

    // Skip end token `]`
    this.next()

    return {
      type: AstNodeType.Array,
      elements,
      location: {
        start: startLocation,
        end: endLocation
      }
    }
  }

  parseElement (): AstElementNode {
    let count: AstElementNode['count']
    let startLocation: TokenLocation

    if (!this.is(TokenChar.countStartToken)) {
      count = 1
    } else {
      startLocation = shallowClone(this.current().location.start)

      // Skip leading count token `(`<-$count)
      this.next()

      const token = this.current()

      if (token.type === TokenType.countNumber) {
        count = token.value as number
        this.next()
      } else {
        count = this.parseIdentifier()

        if (!count) {
          this.throwUnexpected([TokenType.count])
        }
      }

      this.expect(TokenChar.countEndToken)

      // Skip trailing count token ($count->`)`
      this.next()
    }

    const value: AstElementNode['value'] = this.parseValue()

    if (!value) {
      this.throwUnexpectedValue()
    }

    return {
      type: AstNodeType.Element,
      count,
      value,
      location: {
        start: startLocation ?? shallowClone(value.location.start),
        end: shallowClone(value.location.end)
      }
    }
  }

  parseObject (): AstObjectNode {
    if (!this.is(TokenChar.objectStartToken)) {
      return null
    }

    const startLocation: TokenLocation = shallowClone(this.current().location.start)
    const properties: AstObjectNode['properties'] = []

    // Skip start token `{`
    this.next()

    while (this.current() && !this.is(TokenChar.objectEndToken)) {
      properties.push(
        this.parseSpread() ??
        this.parseProperty()
      )
    }

    expect(TokenChar.objectEndToken)

    const endLocation = shallowClone(this.current().location.end)

    // Skip end token `}`
    this.next()

    return {
      type: AstNodeType.Object,
      properties,
      location: {
        start: startLocation,
        end: endLocation
      }
    }
  }

  parseProperty (): AstPropertyNode {
    const key = this.parseIdentifier()

    if (!key) {
      this.throwUnexpectedIdentifier()
    }

    let value: AstPropertyNode['value']

    if (!this.is(TokenChar.objectPairSeparator)) {
      // Copy key node
      value = deepClone(key)
    } else {
      // Skip `:`
      this.next()

      if (!(value = this.parseValue())) {
        this.throwUnexpectedValue()
      }
    }

    return {
      type: AstNodeType.Property,
      key,
      value,
      location: {
        start: shallowClone(key.location.start),
        end: shallowClone(value.location.end)
      }
    }
  }

  parseSpread (): AstSpreadNode {
    const spreadToken = this.current()

    if (spreadToken.type !== TokenType.spread) {
      return null
    }

    // Skip spread token
    this.next()

    const identifier = this.parseIdentifier()

    if (!identifier) {
      this.throwUnexpectedIdentifier()
    }

    return {
      type: AstNodeType.Spread,
      identifier,
      location: {
        start: shallowClone(spreadToken.location.start),
        end: shallowClone(identifier.location.end)
      }
    }
  }

  parseTransform (transformer: AstIdentifierNode): AstIdentifierNode | AstTransformNode {
    if (!transformer || !this.is(TokenChar.transformToken)) {
      return transformer
    }

    // Skip transform token
    this.next()

    const value = this.parseValue()

    if (!value) {
      this.throwUnexpectedValue()
    }

    return {
      type: AstNodeType.Transform,
      transformer,
      value,
      location: {
        start: shallowClone(transformer.location.start),
        end: shallowClone(value.location.end)
      }
    }
  }

  parseIdentifier (): AstIdentifierNode {
    const identifier = this.current()

    if (identifier.type !== TokenType.identifier) {
      return null
    }

    // Advance to next token
    this.next()

    return {
      type: AstNodeType.Identifier,
      name: (identifier.value as string).replace(TRIM_IDENTIFIER_REGEX, ''),
      location: deepClone(identifier.location)
    }
  }

  parseValue (): AstValueNode {
    return (
      this.parseArray() ??
      this.parseObject() ??
      this.parseTransform(this.parseIdentifier())
    )
  }

  expect (token: TokenChar) {
    if (this.current().type === TokenType.EOF) {
      throw new ParserError('Unexpected EOF')
    }

    if (!this.is(token)) {
      this.throwUnexpected([
        token
      ])
    }
  }

  throwUnexpectedIdentifier () {
    this.throwUnexpected([
      TokenType.identifier
    ])
  }

  throwUnexpectedValue () {
    this.throwUnexpected([
      TokenChar.arrayStartToken,
      TokenChar.objectStartToken,
      TokenType.identifier
    ])
  }

  throwUnexpected (expected: string[]) {
    const token = this.current()

    this.throwWithCodeFrame(
      `Unexpected ${token.type === TokenType.EOF ? 'EOF' : `token: ${token.value}`}. Expecting ${expected.join(', ')}`
    )
  }

  throwWithCodeFrame (message: string) {
    throw new ParserError(
      [
        null,
        `${generateCodeframe(this.tokens, this.index)}`,
        message
      ].join('\n\n')
    )
  }

  is (value: string) {
    return this.current().value === value
  }

  current () {
    return this.tokens[this.index]
  }

  next () {
    ++this.index

    return this.current()
  }
}

export function parse (tokens: Token[]) {
  return new Parser(tokens).parse()
}
