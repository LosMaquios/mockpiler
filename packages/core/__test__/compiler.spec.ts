import { createCompiler as mock, raw } from '../src/compiler'
import {
  validContext,
  validCode
} from './fixtures'

describe('compiler', () => {
  test('should compile valid code', () => {
    expect(
      mock(validContext)`${raw(validCode)}`
    ).toMatchSnapshot()
  })

  test('should compile interpolated code', () => {
    expect(
      mock`
        {
          name: ${'Jhon doe'}
          age: ${30}
          fruits: ${['apple', 'strawberry']}
          favoriteFruit: ${() => 'apple'}
        }
      `
    ).toMatchSnapshot()
  })
})
