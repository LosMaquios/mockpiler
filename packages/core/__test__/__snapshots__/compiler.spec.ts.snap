// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler should compile interpolated code 1`] = `
Object {
  "age": 30,
  "favoriteFruit": "apple",
  "fruits": Array [
    "apple",
    "strawberry",
  ],
  "name": "Jhon doe",
}
`;

exports[`compiler should compile valid code 1`] = `
Object {
  "age": 40,
  "matches": Array [
    Object {
      "age": 40,
      "name": "Test",
    },
    Object {
      "age": 40,
      "name": "Test",
    },
    Object {
      "age": 40,
      "name": "Test",
    },
    Object {
      "age": 40,
      "name": "Test",
    },
    Object {
      "age": 40,
      "name": "Test",
    },
    Object {
      "age": 40,
      "name": "Test",
    },
  ],
  "name": "Test",
}
`;
